packages/dd-trace/src/debugger/devtools_client/snapshot/index.js

```diff
@@ -17,26 +19,44 @@ function returnError () {
   return new Error('Error getting local state')
 }
 
-async function getLocalStateForCallFrame (
-  callFrame,
-  {
-    maxReferenceDepth = DEFAULT_MAX_REFERENCE_DEPTH,
-    maxCollectionSize = DEFAULT_MAX_COLLECTION_SIZE,
-    maxFieldCount = DEFAULT_MAX_FIELD_COUNT,
-    maxLength = DEFAULT_MAX_LENGTH
-  } = {}
-) {
+/**
+ * @typedef {Object} GetLocalStateForCallFrameOptions
+ * @property {number} [maxReferenceDepth=DEFAULT_MAX_REFERENCE_DEPTH] - The maximum depth of the object to traverse
+ * @property {number} [maxCollectionSize=DEFAULT_MAX_COLLECTION_SIZE] - The maximum size of a collection to include
+ *   in the snapshot
+ * @property {number} [maxFieldCount=DEFAULT_MAX_FIELD_COUNT] - The maximum number of properties on an object to
+ *   include in the snapshot
+ * @property {number} [maxLength=DEFAULT_MAX_LENGTH] - The maximum length of a string to include in the snapshot
+ * @property {bigint} [deadlineNs=BIGINT_MAX] - The deadline in nanoseconds compared to
+ *   `process.hrtime.bigint()`. If the deadline is reached, the snapshot will be truncated.
+ * @property {boolean} [deadlineReached=false] - Whether the deadline has been reached. Should not be set by the
+ *   caller, but is used to signal the deadline overrun to the caller.
+ */
+
+/**
+ * Get the local state for a call frame.
+ *
+ * @param {import('inspector').Debugger.CallFrame} callFrame - The call frame to get the local state for
+ * @param {GetLocalStateForCallFrameOptions} [opts={}] - The options for the snapshot
+ * @returns {Promise<Object>} The local state for the call frame
+ */
+async function getLocalStateForCallFrame (callFrame, opts = {}) {
+  opts.maxReferenceDepth ??= DEFAULT_MAX_REFERENCE_DEPTH
+  opts.maxCollectionSize ??= DEFAULT_MAX_COLLECTION_SIZE
+  opts.maxFieldCount ??= DEFAULT_MAX_FIELD_COUNT
+  opts.maxLength ??= DEFAULT_MAX_LENGTH
+  opts.deadlineNs ??= BIGINT_MAX
+  opts.deadlineReached ??= false
   const rawState = []
   let processedState = null
 
   try {
-    await Promise.all(callFrame.scopeChain.map(async (scope) => {
-      if (scope.type === 'global') return // The global scope is too noisy
-      rawState.push(...await getRuntimeObject(
-        scope.object.objectId,
-        { maxReferenceDepth, maxCollectionSize, maxFieldCount }
-      ))
-    }))
+    for (const scope of callFrame.scopeChain) {
+      if (opts.deadlineReached === true) break // TODO: Variables in scope are silently dropped: Not the best UX
+      if (scope.type === 'global') continue // The global scope is too noisy
+      // eslint-disable-next-line no-await-in-loop
+      rawState.push(...await getRuntimeObject(scope.object.objectId, opts))
```

Comment:
Is there a reason that this should be run sequential? If yes, I think it would be good to add a comment why.

Reply:
Yes, otherwise we'll queue up a lot of CDP requests that all need to complete even if deadline has been reached, which means we risk overshooting the deadline significantly. Plus, there's no performance improvements gained by running it in parallel since the Chrome DevTools Protocol pushes everything over a serial socket anyway. That's also why I changed the `Promise.all` code inside of the `collector.js` file to only ever run two CDP requests in parallel. Anything higher than that would not have a measurable performance improvement. Anything lower would however take twice as long to run. I assume that's because there's some savings to be gained internally in CDP if a little work is queued up.